### ISSUE TITLE
Leave animation not running in a nested animation of a child component

### ISSUE THREAD

**Issue by marcorinck** [angular/angular#15753]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
I have 2 components (1 parent, 1 child) which should be running 2 different animations at the same time when something triggers in the parent component. The child component is adding an element to the DOM in this case and the enter animation for this element is running fine. When the animation is running backwards because the condition is not met anymore the leave animation in the child component is not running at all. 

When I remove the animation on the parent component the animations of the child component are running fine. 

The animations of the parent component are always running fine. 

This happens only in angular4. Its working as expected in 2.4.10 without any changes.

**Minimal reproduction of the problem with instructions**
Here is the plunkr to demonstrate the problem: https://plnkr.co/edit/a250XgG9FFomAeAYWOSo?p=preview

The yellow block should be moving to the left when leaving, but its not. When you remove "[@move]="direction"" on the div in the parent component you will see that the leave animation is defined correctly because now its running. The enter animation is always fine.

* **Angular version:** 4.0.1 (its working in 2.4.10)

* **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]
all

**Comment by nicky-lenaers** [comment]
Same issue

**Comment by codethief** [comment]
I'm experiencing a similar (the same?) issue:

I have a parent component that has an `expanded` property that is toggled upon click. If `expanded` is `true` a sub-component is added to the view via `*ngIf`. Now sub-component has an animation in its host bindings that is triggered upong enter/leave, i.e.:

```
@Component({
  selector: 'sub-component',
  host: {
    '[style.display]': '"block"',
    '[style.overflow]': '"hidden"',
    '[@enterLeaveAnimation]': '"foo"'
  },

  animations: [
    trigger('enterLeaveAnimation', [
      transition(':leave', [
        style({ height: '*' }),
        animate(250, style({ height: 0 }))
      ]),
      transition(':enter', [
        style({ height: 0 }),
        animate(250, style({ height: '*' }))
      ])
    ])
  ],

  template: `<ng-content></ng-content>`
})
export class SubComponent { }
```

This works fine if the sub-component is a ViewChild of the parent component:

```
<!-- Parent component's template -->
<sub-component *ngIf="expanded">Foo</sub-component>
``` 

However, it does *not* work if the sub-component is actually a ContentChild of the parent component (i.e. the grandparent component's template is something like `<parent-component><sub-component>Foo</subcomponent></parent-component>`) and the parent component's template, at the same time, uses transclusion to insert the sub-component, like so:

```
<!-- Parent component's template -->
<ng-container *ngIf="expanded">
  <ng-content select="sub-component"></ng-content>
</ng-container>
``` 

In this case, the sub-component's animation is not triggered at all.

**Comment by shreenil** [comment]
I'm experiencing the same issue. 
Any updates??

**Comment by sandangel** [comment]
actually there is a PR to fix this issue but it has been revert several times. Hope this will be fix in 5.0.0-rc.3

**Comment by matsko** [comment]
Sorry guys that this fix (#19455) is taking forever to land, but it will fix the issue. Hopefully this week.

**Comment by lares83** [comment]
any update of this issue?

**Comment by oxmanroman** [comment]
nothing yet?

**Comment by oxmanroman** [comment]
Anything new regarding this issue?

**Comment by thw0rted** [comment]
I literally just discovered this feature about half an hour ago, so I don't know if it's a good workaround for this issue, but [the docs](https://angular.io/api/animations/animateChild) state that:

> Each time an animation is triggered in angular, the parent animation will always get priority and any child animations will be blocked

Is this issue related?  I had an animation that looked more or less like 

    transition(":leave", [
        style({transform: "scaleY(1)", height: "*"}),
        animate(transformTiming, style({transform: "scaleY(0)", height: "0"}))
    ])

attached to the parent, and similar animation attached to the child.  When I disabled the parent animation, the child one fired, but if they were both present then the child one never ran.  I added one line to the parent's definition:

    transition(":leave", [
        style({transform: "scaleY(1)", height: "*"}),
        query("@*", [animateChild()], {optional: true}),
        animate(transformTiming, style({transform: "scaleY(0)", height: "0"}))
    ])

Now, it runs exactly as I wanted it to.  On enter, parent grows, then child grows.  On leave, child shrinks, then parent shrinks.  I beat my head against this for a long time before I discovered `query` and `animateChild`.  Maybe it will help those watching this ticket?

**Comment by waterplea** [comment]
2 years later, any update on this? I have a case where parent in created in *ngFor loop in one library and child is animated and comes from another library. :enter animation is triggered but :leave is not and I cannot even try to workaround above because parent does not have any animation on it.

**Comment by johnc-ftl** [comment]
See [the workaround covered here](https://github.com/angular/angular/issues/15798#issuecomment-305934289).

**Comment by johncrim** [comment]
See [the workaround covered here](https://github.com/angular/angular/issues/15798#issuecomment-305934289).

**Comment by cjosue15** [comment]
> Vea [la solución alternativa cubierta aquí](https://github.com/angular/angular/issues/15798#issuecomment-305934289) .

It's the same @thw0rted 's answer

**Comment by gabrielnvian** [comment]
Still no solution to this?

**Comment by dario-piotrowicz** [comment]
Isn't this currently working as expected? :thinking: :shrug: 

As @thw0rted suggested (https://github.com/angular/angular/issues/15753#issuecomment-376915809), parent animations block their children's and that is a know well documented behavior with a clear and well defined solution (using `animeChild`, plus `group` if the animations need to happen in parallel).

Also the `query` token can be anything like `:enter`, `:leave` or even `@*`, so we are not forced to know the trigger's name of the possible children's animations.

___

Moreover @matsko did merge his bugfix PR, in this issue you can see that he added his comment ( (https://github.com/angular/angular/issues/15753#issuecomment-344023099)) + some labels November 13 2017, and the PR got merged the 28th, he did not mention anything here (maybe it just slipped his mind? :shrug:), but I guess that that PR was supposed to close this issue (there were likely bugs with `query` not allowing the standard solution with `animateChild` to work properly).

___

@waterplea about your last comment (https://github.com/angular/angular/issues/15753#issuecomment-638928230), I think this issue is just specific to parent's animations blocking their children's (unless I misunderstood), so if your parent doesn't have any animations yours may probably be a different issue altogether, could you please provide more info and possibly a minimal reproduction? :slightly_smiling_face: (I would actually suggest opening a new issue and if you think that it may be a duplicate of this one add such information + link in the issue's description)

**Comment by lukeify** [comment]
Hi @dario-piotrowicz 

Apologies for bumping this thread, but I don't think this is working expected at all. I have a similar example, but yet is slightly different—here's an example of a `:leave` animation that should firstly run a stateful child animation prior to being destroyed, but the child animation does not occur, even when using the `query()` and `animateChild()` functions from `@angular/animation`, as given in @thw0rted's example. [I've created a minimal reproduction here](https://angular-ivy-tyy7f2.stackblitz.io).

I see there's a bunch of different issues open about similar issues so I'll hold off on opening a new issue for now with the above test case (although if it is indeed a new issue, I'm happy to open one).

**Comment by dario-piotrowicz** [comment]
Thanks for your input @lukeify :slightly_smiling_face: 

I still stand with my belief that that the original discussion refers to something that is already fixed

___

I've played around and tried debugging stuff and to me your bug seems to be something different and I am not sure if that is actually a bug and/or anything can be done about it.


The issue in your example is not that the `query`/`animateChild` combo doesn't work, the issue is instead that the child component is not transitioning from `organgered` to `dodgerblue` but from `orangered` to `void`.

As you can see from the console mid-animation:
![Screenshot at 2022-01-27 20-48-05](https://user-images.githubusercontent.com/61631103/151443805-3839f4ff-3d66-4afa-b8e3-67717e8ae115.png)


This happens because the element is leaving the DOM as part of its parent removal (tidbit, the 
[_signalRemovalForInnerTriggers](https://github.com/angular/angular/blob/b66e479cdb1e474a29ff676f10a5fcc3d7eae799/packages/animations/browser/src/render/transition_animation_engine.ts#L319) is specifically setting the element in a leaving state before the `animateChild` call applies).

So your animation does not break because of `query` and/or `animateChild` but because the element is seen as leaving the DOM and not transitioning from a state to another, as you can for example see here: [original stackblitz + leave transition for child](https://stackblitz.com/edit/angular-ivy-chfj2h?file=src%2Fapp%2Fchild%2Fchild.component.ts) if you cater for that case (which you currently don't) your animation does work as it should.

As I said anyways the child element is indeed leaving the DOM and that seems to get precedence over the fact that you're changing the trigger's value, I am not sure, I think this may probably be the intended behavior and we as developers need to adapt to such (by for example adding the `:leave` transition as I mentioned).

You may disagree and think that it is a bug, if that is the case I would suggest you to open a new issue :slightly_smiling_face:

**Comment by thePunderWoman** [comment]
This issue was closed as obsolete due to the introduction of the new `animate.enter` and `animate.leave` APIs. The `@angular/animations` package is now deprecated and will be removed in Angular v23. We hope you give the new APIs a try.
